README.md
# :earth_americas: Orbital Decay Simulator
An interactive web app that allows for approximating and visualizing a spacecraft's orbital decay and reentry.

Check the app out [here](https://orbital-decay-simulator.herokuapp.com/).

This simulator works by calculating how drag force affects orbital elements over
time. More complicated causes of orbital decay, such as solar radiation pressure, are not included as part
of the decay calculations. The 1976 U.S. Standard Atmosphere model is used to calculate the density of the atmosphere at a
particular altitude.

Start by supplying a few orbital elements and viewing the accompanying visualization. Next, supply some information about your
spacecraft. Finally, click the "Simulate" button to run the simulation. When the simulation is completed, you will be given
an estimate of when the spacecraft will land back on Earth as well as visualizations of the vehicle's telemetry while its orbit decayed.

## Technologies Used
#### [Streamlit](https://streamlit.io/)
- Front-end framework designed for data scientists and machine learning developers to quickly add interfaces to their projects

#### [Plotly](https://plotly.com/)
- A data visualization library that allows for interactive 2D and 3D plots

#### [poliastro](https://docs.poliastro.space/en/stable/)
- Python library for visualizing and calculating orbits and orbital manuevers. Used to plot ground tracks

#### [Astropy](https://www.astropy.org/)
- Python library for general astronomy calculations

#### [Altair](https://altair-viz.github.io/)
- A declarative data visualization library for Python

#### [Pandas](https://pandas.pydata.org/)
- A data science library for Python used to store and transform data

#### [NumPy](https://numpy.org/)
- A data science library for Python that supplies a powerful array data structure often preferable to Python's built-in list structure

## References
- [National Oceanic and Atmospheric Administration, National Aeronautics and Space Administration, United States Air Force, *U.S. Standard Atmosphere, 1976*](https://www.ngdc.noaa.gov/stp/space-weather/online-publications/miscellaneous/us-standard-atmosphere-1976/us-standard-atmosphere_st76-1562_noaa.pdf)
- [R. A. Braeunig, *Atmospheric Models*](http://www.braeunig.us/space/atmmodel.htm)
- [USAFA Astronautics and Space Ops, *LSN 11 - Ground Tracks* [YouTube]](https://youtu.be/Q4AQT0vDV_M)
- R. R. Bate, D. D. Mueller, J. E. White, and W. W. Saylor, *Fundamentals of Astrodynamics.* Second Edition


changelog.md
# Changelog
### v1.1
- Improved ground track plots by using poliastro's built-in `GroundPlotter`
- Improved possible landing locations map by incorporating new ground track plotting function
- Added additional optional orbital element inputs to improve accuracy of plotted ground tracks
- Added option to change ground track map projections 

src/model.py
# File containing functions needed to parse .obj files and deal with 3D spacecraft models

import numpy as np
import pandas as pd
from io import StringIO
from scipy.spatial import ConvexHull
import streamlit as st

# Symbol that indicates that current line represents a vertex
_VERTEX_RECORD = "v"

# Index that holds record symbol ("v", "l", etc.)
_RECORD_INDEX = 0

# Indices for vector info in a vector record
_OBJ_X = 1
_OBJ_Y = 2
_OBJ_Z = 3

# Indices for coordinate info in a set of Numpy arrays of coordinates
_ARRAY_X = 0
_ARRAY_Z = 1

# Parses .obj file given Streamlit UploadedFile object and returns a Pandas dataframe with the
# X, Y, and Z coordinates of its vertices
@st.cache
def parseObj(obj):
	# Create arrays to hold x, y, and z vertex coordinates
	x = np.empty(0)
	y = np.empty(0)
	z = np.empty(0)

	# Get file contents from Streamlit file uploader
	objData = StringIO(obj.getvalue().decode("UTF-8")).read().split("\n")

	# Iterate through .obj to find vertex and line info
	for line in objData:
		# Strip whitespace from line and split at inner spaces
		lineSplit = line.strip().split(" ")
		
		# If this line contains vertex info
		if lineSplit[_RECORD_INDEX] == _VERTEX_RECORD:
			x = np.append(x, float(lineSplit[_OBJ_X]))
			y = np.append(y, float(lineSplit[_OBJ_Y]))
			z = np.append(z, float(lineSplit[_OBJ_Z]))

	return pd.DataFrame({"x" : x, "y" : y, "z" : z})

# Takes a X, Y, and Z coordinates to project onto the XZ plane (the plane normal to the velocity vector).
# Returns the projected coordinates
def _projectPoints(x, y, z):
	y *= 0
	return (x, y, z)

# Takes the X, Y, and Z coordinates of a 3D object and return the area of the orthographic projection
# in m^2
def calculateOrthographicProjectionArea(x, y, z):
	# Find all unique outer points from the projection
	outerPoints = ConvexHull(np.stack((x, z), -1))

	return outerPoints.volume


src/plotting.py
# File containing functions used to plot visualizations
from orbital_mechanics import *
import plotly.graph_objects as go
import streamlit as st
import altair as alt
from poliastro.earth.plotting import GroundtrackPlotter
from poliastro.earth import EarthSatellite
from poliastro.twobody import Orbit
from poliastro.bodies import Earth
from astropy import units as u
from poliastro.util import time_range
import plotly.graph_objects as go

# Color of ground track points
_TRACK_COLOR = "darkblue"

# Map element colors
_LAND_COLOR = "white"
_WATER_COLOR = "rgb(140, 181, 245)"
_COUNTRY_COLOR = "lightgray"

# Color of launch site marker
_LAUNCH_SITE_COLOR = "crimson"

# Plotly projection types
PROJECTION_TYPES = [
	"Equirectangular",
	"Mercator",
	"Orthographic",
	"Natural Earth",
	"Kavrayskiy7",
	"Miller",
	"Robinson",
	"Eckert4",
	"Azimuthal Equal Area",
	"Azimuthal Equidistant",
	"Conic Equal Area",
	"Conic Conformal",
	"Conic Equidistant",
	"Gnomonic",
	"Stereographic",
	"Mollweide",
	"Hammer",
	"Transverse Mercator",
	"Winkel Tripel",
	"Aitoff",
	"Sinusoidal"
]

# Returns Plotly figure of scatter mapbox plot of the ground track of an orbit given: apogee, perigee,
# inclination, right ascension of the ascending node, argument of perigee, true anomaly, starting lat,
# and starting lon
@st.cache(show_spinner = False)
def plotGroundTrack(a, p, i, raan, argOfPerigee, trueAnomaly, startingLat, startingLon, proj = PROJECTION_TYPES[0], showMany = False):
	fig = GroundtrackPlotter()

	# Derive orbital elements needed to plot ground track
	period = calculateOrbitalPeriod(a, p)
	semiMajoraxis = calculateSemiMajorAxis(a, p)
	eccentricity = calculateEccentricity(a, p)

	# Multiply period if many orbit tracks are to be shown
	if showMany:
		period *= 86400 / period / 2

	# Create objection needed for poliastro GroundPlotter
	orbit = Orbit.from_classical(Earth, semiMajoraxis * u.km, eccentricity * u.one, i * u.deg, raan * u.deg, argOfPerigee * u.deg, trueAnomaly * u.deg)
	sat = EarthSatellite(orbit, None)
	t_span = time_range(orbit.epoch - period * u.s, periods = 150, end = orbit.epoch + period * u.s)
	
	# Plot ground track
	fig.plot(
		sat,
		t_span,
		label = "Ground Track",
		color = _TRACK_COLOR,
		marker={"size": 0, "symbol": "triangle-right"},
	)

	# Format map
	fig.fig.update_geos(
		bgcolor = "rgba(0, 0, 0, 0)",
		showframe = False,
		lataxis = {"showgrid" : False},
		lonaxis = {"showgrid" : False},
		showlakes = True,
		showcountries = True,
		showrivers = True,
		oceancolor = _WATER_COLOR,
		landcolor = _LAND_COLOR,
		lakecolor = _WATER_COLOR,
		rivercolor = _WATER_COLOR,
		countrycolor = _COUNTRY_COLOR,
		projection_type = proj.lower()
	)

	fig.fig.update_traces(
		marker = {"color" : "rgba(0, 0, 0, 0)"},
	)

	fig.fig.update_layout(
		showlegend = False,
		margin = {"l" : 0, "r" : 0, "b" : 0, "t" : 0}
	)

	# Add launch site point
	fig.add_trace(
		go.Scattergeo(
			lat = [startingLat],
			lon = [startingLon],
			name = "Launch Site 🚀",
			marker = {"color" : _LAUNCH_SITE_COLOR}
		)
	)

	return fig.fig

# Returns Altair chart visualizing telemetry data. Takes telemetry dataframe, name of column to visualize over time, and the
# Y-axis label for the plot
def plotTelemetry(telemetry, column, label):
	# Get smallest and largest datapoints for scaling
	smallestY = telemetry[column].min()
	largestY = telemetry[column].max()

	smallestX = telemetry["time"].min()
	largestX = telemetry["time"].max()

	return alt.Chart(telemetry).mark_line().encode(
		x = alt.X("time", axis = alt.Axis(title = "Time (s)"), scale = alt.Scale(domain = (smallestX, largestX))),
		y = alt.Y(column, axis = alt.Axis(title = label), scale = alt.Scale(domain = (smallestY, largestY))),
		tooltip = ["time", column]
	).add_selection(
		alt.selection_interval(bind = "scales")
	)

src/atmosphere.py
# File containing implementation of 1976 U.S. Standard Atmosphere

import math

# Universal gas constant in N*km/(kmol*K)
_R = .00831432

# Acceleration due to gravity at sea level in km/s^2
_G0 = .00980665

# Effective radius of the Earth in km
_R0 = 6356.766

# Mean molecular weight of gasses at sea level in kg/kmol
_M0 = 28.9644

# Geopotential height reference at various subscripts. Corresponds
# to Table 4 in the 1976 U.S. Standard Atmosphere paper
_REF_GEOPOTENTIAL_HEIGHT = [0, 11, 20, 32, 47, 51, 71, 84.8520]

# Molecular-scale temperature gradient in K/km' at various subscripts. Corresponds
# to Table 4 in the 1976 U.S. Standard Atmosphere paper
_MOL_SCALE_TEMP_GRADIENT = [-6.5, 0, 1.0, 2.8, 0, -2.8, -2.0]

# Molecular-scale temperature in K at various subscripts. Calculated using
# the given sea level temperature and the molecular-scale temperature
# gradients at each height
_REF_MOL_SCALE_TEMP = [288.15, 216.65, 216.65, 228.65, 270.65, 270.65, 214.65, 186.87]

# Reference pressures for various subscripts. Calculated using the pressure equations in the U.S.
# Standard Atmosphere paper. All pressures are in Pa
_REF_PRESSURE = [101325, 22632.06, 5474.89, 868.02, 110.91, 66.94, 3.96]

# Coefficients for a polynomial needed for solve for the density above 86 km.
# Coefficients come from Robert Braeunig: www.braeunig.us/space/atmmodel.htm
_DENSITY_COEFFICIENTS = [
	(0, -3.322622e-6, 9.111460e-4, -0.2609971, 5.944694),
	(0, 2.873405e-5, -0.008492037, 0.6541179, -23.62010),
	(-1.240774e-5, 0.005162063, -0.8048342, 55.55996, -1443.338),
	(0, -8.854164e-5, 0.03373254, -4.390837, 176.5294),
	(3.661771e-7, -2.154344e-04, 0.04809214, -4.884744, 172.3597),
	(1.906032e-8, -1.527799e-05, 0.004724294, -0.6992340, 20.50921),
	(1.199282e-9, -1.451051e-6, 6.910474e-4, -0.1736220, -5.321644),
	(1.140564e-10, -2.130756e-7, 1.570762e-4, -0.07029296, -12.89844),
	(8.105631e-12, -2.358417e-9, -2.635110e-6, -0.01562608, -20.02246),
	(-3.701195e-12, -8.608611e-9, 5.118829e-5, -0.06600998, -6.137674)
]

# Raises exception when an input is out of bounds of the Standard Atmosphere height range
def _checkInStandardAtmosphereRange(z):
	if z < 0 or z > 1000:
		raise Exception("Geometric height must be between 0 and 1000 km")

# Raises exception when an input is not within the 0 to 86 km height range
def _checkBelow86(z):
	if z < 0 or z > 86:
		raise Exception("Geometric height must be between 0 and 86 km")

# Raises exception when an input is not within the 86 (non-inclusive) to 1000 km range
def _checkAbove86(z):
	if z <= 86 or z > 1000:
		raise Exception("Geometric height must be between 87 and 1000 km")

# Converts the given geometric height in km to geopotential km'
def _geometricToGeopotentialHeight(z):
	_checkInStandardAtmosphereRange(z)

	return (_R0 * z) / (_R0 + z)

# Converts the given geometric height to a subscript between 0 and 6 as defined in Table 4 of the U.S.
# Standard Atmosphere paper
def _heightToSubscriptBelow86(z):
	_checkBelow86(z)

	# Convert to km'
	h = _geometricToGeopotentialHeight(z)

	# Check various ranges and return correct subscript
	if h >= 0 and h <= 11:
		return 0
	elif h > 11 and h <= 20:
		return 1
	elif h > 20 and h <= 32:
		return 2
	elif h > 32 and h <= 47:
		return 3
	elif h > 47 and h <= 51:
		return 4
	elif h > 51 and h <= 71:
		return 5
	elif (h > 71 and h <= 84.852) or math.isclose(h, 84.852, rel_tol = 1e-6):
		return 6

# Converts the given geometric height to a subscript between 0 and 9 line up with altitude ranges
# in Table 8 of Robert Braeunig's page on atmospheric models
def _heightToSubscriptAbove86(z):
	_checkAbove86(z)
	
	# Check which altitude range the given height falls into
	if z > 86 and z <= 91:
		return 0
	elif z > 91 and z <= 100:
		return 1
	elif z > 100 and z <= 110:
		return 2
	elif z > 110 and z <= 120:
		return 3
	elif z > 120 and z <= 150:
		return 4
	elif z > 150 and z <= 200:
		return 5
	elif z > 200 and z <= 300:
		return 6
	elif z > 300 and z <= 500:
		return 7
	elif z > 500 and z <= 750:
		return 8
	elif z > 750 and z <= 1000:
		return 9

# Returns pressure at given geometric height. Raises exception if z is not between 0 to 86 km.
# Pressure returned is in Pascals (Pa)
def _getPressureBelow86(z):
	# Get subscript corresponding to height
	subscript = _heightToSubscriptBelow86(z)

	# Get geopotential height
	h = _geometricToGeopotentialHeight(z)

	# Get molecular scale temp gradient corresponding to this subscript
	molScaleTempGradient = _MOL_SCALE_TEMP_GRADIENT[subscript]

	# Get reference molecular scale temp corresponding to this subscript
	referenceMolScaleTemp = _REF_MOL_SCALE_TEMP[subscript]

	# Get pressure corresponding to this subscript
	referencePressure = _REF_PRESSURE[subscript]

	# Get reference geopotential altitude corresponding to this subscript
	referenceHeight = _REF_GEOPOTENTIAL_HEIGHT[subscript]

	# Decide which pressure equation to use
	if molScaleTempGradient != 0:
		# Calculate exponent
		exponent = (_G0 * _M0) / (_R * molScaleTempGradient)

		# Calculate value notated in brackets in paper
		bracketValue = referenceMolScaleTemp / (referenceMolScaleTemp + molScaleTempGradient * (h - referenceHeight))

		return referencePressure * (bracketValue**exponent)
	else:
		# Calculate value notated in brackets in paper
		bracketValue = ((-_G0 * _M0) * (h - referenceHeight)) / (_R * referenceMolScaleTemp)

		return referencePressure * math.exp(bracketValue)

# Returns molecular-scale temp below 86 km. Raises exception if z is not between 0 to 86 km
def _getMolScaleTempBelow86(z):
	_checkBelow86(z)

	# Get subscript corresponding to this altitude
	subscript = _heightToSubscriptBelow86(z)

	# Convert geometric to geopotential height
	h = _geometricToGeopotentialHeight(z)

	return _REF_MOL_SCALE_TEMP[subscript] + _MOL_SCALE_TEMP_GRADIENT[subscript] * (h - _REF_GEOPOTENTIAL_HEIGHT[subscript])

# Returns density at given geometric height. Raises exception if z is not between 0 to 86 km.
# Returns in units of kh/km^3 to match distance units used everywhere else in the program
def _getDensityBelow86(z):
	_checkBelow86(z)

	# Get subscript at given altitude
	subscript = _heightToSubscriptBelow86(z)

	# Get pressure at given altitude. Convert so distance units match with other values
	pressure = _getPressureBelow86(z) * 1000

	return (pressure * _M0) / (_R * _getMolScaleTempBelow86(z))

# Returns density at given geometric height. Raises exception if z is not between 87 to 1000 km
def _getDensityAbove86(z):
	_checkAbove86(z)

	# Get subscript corresponding given altitude
	subscript = _heightToSubscriptAbove86(z)

	# Get coefficients corresponding to this altitude range
	a, b, c, d, e = _DENSITY_COEFFICIENTS[subscript]

	# Return polynomial calculated with correct coefficients
	return math.exp(a * z**4 + b * z**3 + c * z**2 + d * z + e) * 1000000000

# Returns density at given geometric height. Raises exception if z is not between 0 to 1000 km.
# Density unit it kg/km^3
def getDensity(z):
	_checkInStandardAtmosphereRange(z)

	# Check range and call appropriate function
	if z <= 86:
		return _getDensityBelow86(z)
	else:
		return _getDensityAbove86(z)

src/orbital_mechanics.py
# File containing functions used for calculating orbital elements
import numpy as np
import pandas as pd
import streamlit as st
from atmosphere import *

# Standard gravitational parameter of Earth in km^3/s^2
MU = 3.986004e5

# Earth radius in km
RADIUS = 6371

# Max number of points that the simulation will return in a dataframe
_MAX_POINTS = 200000

# Number of degrees Earth turns in a second
_OMEGA_EARTH = 0.00417806712

# J2 constant for Earth
J2 = 1.08262668e-3

# Throws error if given apogee is < given perigee
def _checkExtrema(a, p):
	if a < p:
		raise Exception("Apogee must be larger than or equal to perigee")

# Returns the orbital period in seconds given the apogee and perigee in km
def calculateOrbitalPeriod(a, p):
	_checkExtrema(a, p)

	semiMajorAxis = calculateSemiMajorAxis(a, p)

	return 2 * math.pi * math.sqrt(semiMajorAxis**3 / MU)

# Returns semi-major axis in km from apogee and perigee in km
def calculateSemiMajorAxis(a, p):
	_checkExtrema(a, p)
	return ((a + RADIUS) + (p + RADIUS)) / 2

# Returns eccentricity of orbit from apogee and perigee in km
def calculateEccentricity(a, p):
	_checkExtrema(a, p)
	return ((a + RADIUS) - (p + RADIUS)) / (2 * calculateSemiMajorAxis(a, p))

# Returns the angular velocity given an orbital velocity and distance from body
def calculateAngularVelocity(v, r):
	return v / r

# Returns result of Kepler's first law given apogee, perigee, and the number of degrees from perigee
def calculateMainFocusDistance(a, p, theta):
	_checkExtrema(a, p)

	# Calculate other elements needed
	semiMajorAxis = calculateSemiMajorAxis(a, p)
	eccentricity = calculateEccentricity(a, p)

	# Calculate numerator and denominator separately of the equation represented by Kepler's first law
	numerator = semiMajorAxis * (1 - eccentricity**2)
	denominator = 1 + eccentricity * np.cos(np.radians(theta))

	return numerator / denominator

# Calculates orbital velocity at a given angle theta from perigee given apogee and perigee
def calculateOrbitalVelocity(a, p, theta):
	_checkExtrema(a, p)

	# Calculate other elements needed
	semiMajorAxis = calculateSemiMajorAxis(a, p)
	distance = calculateMainFocusDistance(a, p, theta)

	ratioDifference = (2 / distance) - (1 / semiMajorAxis)

	return np.sqrt(MU * ratioDifference)

# Calculates the semi-major axis from the Vis-viva equation. Used to determine how the semi-major
# axis of an orbit changes due to velocity changes caused by drag force. Takes the distance from
# the body being orbited and the instantaneous orbital velocity
def calculateSemiMajorAxisFromVisViva(r, v):
	return -(MU * r) / (v**2 * r - 2 * MU)

# Returns the acceleration experienced by the given mass at the given height with the given velocity,
# drag coefficient, and reference area. Note that this uses Newton's second law F = ma and does
# not take into account any effects of the velocity approaching the speed of light
@st.cache(show_spinner = False)
def calculateAccelerationFromDrag(m, z, v, cd, area):
	# Check for non-zero mass
	if m == 0:
		raise Exception("Mass cannot be 0")

	density = getDensity(z)

	# Calculate drag force
	dragForce = .5 * density * v**2 * cd * area

	# Return acceleration
	return dragForce / m

# Main driver for orbital decay simulation. Takes initial apogee, perigee, inclination, drag coefficient, average
# cross-sectional area in m^2, and the time step for the simulation.
@st.cache(show_spinner = False)
def simulateOrbitalDecay(a, p, i, m, cd, area, timeStep):
	# Convert area to km^2
	area *= 1e-6

	# Initial parameters
	theta = 0
	time = 0
	distance = calculateMainFocusDistance(a, p, theta)
	altitude = distance - RADIUS
	eccentricity = calculateEccentricity(a, p)
	semiMajorAxis = calculateSemiMajorAxis(a, p)

	telemetry = {"time" : [], "dragAcceleration" : [], "velocity" : [], "apogee" : [], "perigee" : []}

	# Main simulation loop
	while altitude >= 0:
		distance = calculateMainFocusDistance(a, p, theta)
		velocity = calculateOrbitalVelocity(a, p, theta)
		altitude = distance - RADIUS

		if altitude <= 0 or altitude > 1000:
			break

		# Update velocity based on the acceleration due to atmospheric drag
		dragAcceleration = calculateAccelerationFromDrag(m, altitude, velocity, cd, area)
		velocity -=  timeStep * dragAcceleration
		
		theta +=  timeStep * calculateAngularVelocity(velocity, distance)
		semiMajorAxis = calculateSemiMajorAxisFromVisViva(distance, velocity)
		
		# Recalculate apogee and perigee for next iteration
		a = semiMajorAxis * (1 + eccentricity) - RADIUS
		p = semiMajorAxis * (1 - eccentricity) - RADIUS

		if a <= 0 or p <= 0 or a < p:
			break

		telemetry["time"].append(time)
		telemetry["dragAcceleration"].append(dragAcceleration)
		telemetry["velocity"].append(velocity)
		telemetry["apogee"].append(a)
		telemetry["perigee"].append(p)

		time += timeStep

	# Make sure apogee and perigee aren't negative
	if a < 0:
		a = 0
	if p < 0:
		p = 0

	# Try to create dataframe from telemetry if there is not too much data
	telemetryDataFrame = None

	try:
		telemetryDataFrame = pd.DataFrame(telemetry)
	
		# Even if dataframe is within size limits, too many records could slow down browser
		if len(telemetryDataFrame) > _MAX_POINTS:
			telemetryDataFrame = None
	except:
		pass
	
	return (time, telemetryDataFrame)

src/interface.py
# File containing Streamlit interface controls
import streamlit as st
from plotting import *
from datetime import timedelta, datetime, time

# Format string for degree inputs
_DEGREE_FORMAT = "%d°"

st.set_page_config(page_title = "Orbital Decay Simulator", layout = "centered", page_icon = "../assets/favicon.png")

# Main driver for Streamlit inputs and calling of other files' functions
def main():
	# Blank string needed to ensure Streamlit recognizes first multi-line string as markdown
	""

	# Intro text
	"""
	# :earth_americas: Orbital Decay Simulator
	Calculates and visualizes orbital decay.
	Written by [Jack Sheehan](https://github.com/JackCSheehan).

	This tool is open source. Find a bug or want to contribute? Check out the source on
	[GitHub](https://github.com/JackCSheehan/orbital-decay-simulator).

	**Disclaimer:** This tool is for educational/hobbyist purposes only. Mistakes in calculations
	will be fixed as soon as possible, but this tools uses generalized and approximated models. **Real-life
	accuracy should not be expected.**

	---
	## How it Works
	This simulator works by calculating how drag force affects orbital elements over
	time. More complicated causes of orbital decay, such as solar radiation pressure, are not included as part
	of the decay calculations. The 1976 U.S. Standard Atmosphere model is used to calculate the density of the atmosphere at a
	particular altitude.

	Start by supplying a few orbital elements and viewing the accompanying visualization. Next, supply some information about your
	spacecraft. Finally, click the "Simulate" button to run the simulation. When the simulation is completed, you will be given
	an estimate of when the spacecraft will land back on Earth as well as visualizations of the vehicle's trajectory while its orbit decayed.
	"""

	"""
	---
	## Part 1: Orbital Parameters
	Enter the basic parameters of your spacecraft's starting orbit.
	
	Only the apogee, perigee, inclination, and insertion date/time need to be provided for the
	simulator to work. The remaining parameters are only needed to more precisely plot the ground track of your orbit. If you do not care about the ground track's
	accuracy, you can leave the RAAN, argument of perigee, and true anomaly fields at their default values.
	
	The launch latitude and longitude are also optional as they simply allow you to visualize the launch site on the map.
	"""

	orbitCol1, orbitCol2 = st.columns(2)

	with orbitCol1:
		# Placeholder for lat input
		latInput = st.empty()

		startingLat = latInput.number_input("Launch Site Latitude (°)", -90.0, 90.0, 0.0, help = "The latitude, in degrees, that your vehicle launched from")
		apogee = st.number_input("Apogee (km)", 150, 1000, help = "The distance the farthest part of your orbit is from Earth's surface")
		startDate = st.date_input("Orbit insertion date", help = "Date of insertion into starting orbit")
		inclination = st.slider("Inclination (°)", 0, 90, 0, format = _DEGREE_FORMAT, help = "The angle of your orbit with respect to Earth's equatorial plane")
		argOfPerigee = st.slider("Argument of Perigee (°)", 0, 360, 0, format = _DEGREE_FORMAT, help = "Angle between the ascending node and perigee")

	with orbitCol2:
		# Placeholder inputs for lon input
		lonInput = st.empty()

		startingLon = lonInput.number_input("Launch Site Longitude (°)", -180.0, 180.0, 0.0, help = "The longitude, in degrees, that your vehicle launched from")
		perigee = st.number_input("Perigee (km)", 150, 1000, help = "The distance the closest part of your orbit is from Earth's surface")
		startTime = st.time_input("Orbit insertion time (UTC)", value = time(0, 0, 0), help = "Time of insertion into starting orbit. Uses 24-hour time")
		raan = st.slider("Right Ascension of the Ascending Node (°)", 0, 360, 0, format = _DEGREE_FORMAT, help = "Angle between the I unit vector and the ascending node")
		trueAnomaly = st.slider("True Anomaly (°)", 0, 360, 0, format = _DEGREE_FORMAT, help = "Angle between spacecraft's position and perigee at epoch")

	# Check that inclination is valid given launch site latitude
	if inclination < abs(startingLat):
		"""
		#### Orbital inclination cannot be less than starting latitude
		"""
		return

	# Check that apogee is >= perigee
	if apogee < perigee:
		"""
		#### Apogee must be greater than or equal to perigee
		"""
		return

	"""
	#### Ground Track
	Shown below is the ground track of your orbit.
	"""
	proj = st.selectbox("Map Projection", PROJECTION_TYPES)
	st.plotly_chart(plotGroundTrack(apogee, perigee, inclination, raan, argOfPerigee, trueAnomaly, startingLat, startingLon, proj), use_container_width = True)

	"""
	---
	## Part 2: Spacecraft and Simulation Parameters
	"""

	"""
	#### Basic Information
	Enter in some basic information about your spacecraft.
	"""

	craftCol1, craftCol2 = st.columns(2)

	with craftCol1:
		mass = st.number_input("Mass (kg)", 1, None, 1000, help = "The constant mass of your spacecraft")

	with craftCol2:
		dragCoefficient = st.number_input("Drag Coefficient", .10, 5.00, 1.15, help = "Also know as coefficient of drag. Often approximated as 2.2 for spherical spacecraft")
	
	averageArea = st.number_input("Average cross-sectional area (m²)", .0001, None, 1.00, help = "The average cross-sectional area of your spacecraft perpendicular to airflow")
	timeStep = st.number_input("Time Step (s)", 1, 3600, 300, help = "The number of seconds skipped in simulation loop. Lower numbers are more accurate, but take much longer. Larger numbers are faster, but lead to less precise visualizations and predictions")

	"""
	---
	## Part 3: Simulation and Visualizations
	"""

	# Show spinner and call appropriate functions
	with st.spinner("Please wait while your simulation is being run..."):
		pass

	if st.button("Simulate"):
		with st.spinner("Running Simulation..."):
			totalElapsedSeconds, telemetry = simulateOrbitalDecay(apogee, perigee, inclination, mass, dragCoefficient, averageArea, timeStep)
			
			startDatetime = datetime.combine(startDate, startTime)

			landingDatetime = startDatetime + timedelta(seconds = totalElapsedSeconds)
			landingDateStr = landingDatetime.strftime("%B %d, %Y at %H:%M:%S %Z")

			f"""
			#### Estimated Landing

			`{landingDateStr}`
			"""

			if inclination == 0:
				"The spacecraft was put into an equatorial orbit and as such can only land on the Equator, which is shown on the map below."
			else:
				f"""
				The spacecraft has the possibility of landing between `{inclination}°` latitude and `-{inclination}°` latitude. The spacecraft's ground track over several orbits is shown on the map below.
				"""
				

			st.plotly_chart(plotGroundTrack(apogee, perigee, inclination, raan, argOfPerigee, trueAnomaly, startingLat, startingLon, proj, True), use_container_width = True)

		if telemetry is None:
			"""
			There are too many data points to generate visualizations. Please increases the time step and rerun the simulation.
			"""
		else:
			with st.spinner("Generating Visualizations..."):
				"""
				#### Telemetry Visualizations

				Left click and drag to pan the plots. Use your mouse wheel to zoom. Hovering your mouse over a point will show its exact values.
				"""

				# Plot telemetry
				"#### Apogee Over Time"
				st.altair_chart(plotTelemetry(telemetry, "apogee", "Apogee (km)"), use_container_width = True)

				"#### Perigee Over Time"
				st.altair_chart(plotTelemetry(telemetry, "perigee", "Perigee (km)"), use_container_width = True)

				"#### Velocity Over Time"
				st.altair_chart(plotTelemetry(telemetry, "velocity", "Velocity (km/s)"), use_container_width = True)

				"#### Acceleration Due to Drag Over Time"
				st.altair_chart(plotTelemetry(telemetry, "dragAcceleration", "Acceleration Due to Drag m/s²"), use_container_width = True)

if __name__ == "__main__":
	main()